File: spec/stack_spec.rb
# frozen_string_literal: true

require_relative '../lib/stack.rb'

RSpec.describe 'Stack - Ruby' do
  stack = Stack.new

  it 'has an `empty?` method' do
    expect(stack).to respond_to(:empty?)
  end

  it 'is empty when newly created' do
    expect(stack.empty?).to be true
  end

  it 'has a `push` method' do
    expect(stack).to respond_to(:push)
  end

  it 'has a `pop` method' do
    expect(stack).to respond_to(:pop)
  end

  it '`pop` method returns the last pushed value' do
    stack = Stack.new

    stack.push(40)
    stack.push(50)
    expect(stack.pop).to eq(50)
  end

  it '`pop` method removes the last value from the store' do
    stack = Stack.new

    stack.push(40)
    stack.push(50)
    stack.pop
    expect(stack.pop).to eq(40)
  end
end

File: spec/stack.spec.js
'use strict'

// Allow chai syntax like `expect(foo).to.be.ok;`
// jshint -W030

const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')

chai.use(chaiAsPromised)

const expect = chai.expect

const Stack = require('../lib/stack')

describe('Stack - JS', function () {
  let stack = new Stack()

  it('has an `isEmpty` method', function () {
    expect(stack).to.respondTo('isEmpty')
  })

  it('is empty when newly created', function () {
    expect(stack.isEmpty()).to.be.true
  })

  it('has a `push` method', function () {
    expect(stack).to.respondTo('push')
  })

  it('has a `pop` method', function () {
    expect(stack).to.respondTo('pop')
  })

  it('`pop` method returns the last pushed value', function () {
    stack = new Stack()

    stack.push(100)
    stack.push(200)
    expect(stack.pop()).to.eq(200)
  })

  it('`pop` method removes the last value from the store', function () {
    stack = new Stack()

    stack.push(100)
    stack.push(200)
    stack.pop()
    expect(stack.pop()).to.eq(100)
    expect(stack.pop()).to.be.undefined
  })
})
